hera_cal/frf.py

```diff
@@ -642,22 +1001,101 @@ def tophat_frfilter(self, keys=None, wgts=None, mode='clean',
+                                                                      percentile_high=percentile_high,
+                                                                      keys=keys, verbose=verbose,
+                                                                      frate_standoff=frate_standoff, fr_freq_skip=fr_freq_skip,
+                                                                      frate_width_multiplier=frate_width_multiplier,
+                                                                      min_frate_half_width=min_frate_half_width)
+
         wgts = io.DataContainer({k: (~self.flags[k]).astype(float) for k in self.flags})
+        if pre_filter_modes_between_lobe_minimum_and_zero:
+            self.pre_filter_resid = DataContainer({})
+            self.pre_filter_resid_flags = DataContainer({})
+            filter_kwargs_no_data = copy.deepcopy(filter_kwargs)
+            if 'data' in filter_kwargs_no_data:
+                del filter_kwargs_no_data['data']
         for k in keys:
             if mode != 'clean':
                 filter_kwargs['suppression_factors'] = [tol]
             else:
                 filter_kwargs['tol'] = tol
-            self.fourier_filter(keys=[k], filter_centers=[center_frates[k]], filter_half_widths=[width_frates[k]],
-                                mode=mode, x=self.times * 3.6 * 24.,
-                                data=self.data, flags=self.flags, wgts=wgts,
-                                ax='time', cache=filter_cache, skip_wgt=skip_wgt, verbose=verbose, **filter_kwargs)
+
+            if pre_filter_modes_between_lobe_minimum_and_zero:
+                min_frate_abs = np.min([np.abs(frate_centers[k] + frate_half_widths[k]),
+                                        np.abs(frate_centers[k] - frate_half_widths[k])])
+                # only pre-filter if main-lobe does not include zero fringe-rates.
+                if not (frate_centers[k] + frate_half_widths[k] >= 0. and frate_centers[k] - frate_half_widths[k] <= 0.):
+                    self.fourier_filter(keys=[k], filter_centers=[0.], filter_half_widths=[min_frate_abs],
+                                        mode=mode, x=self.times * SDAY_KSEC,
+                                        wgts=wgts, output_prefix='pre_filter',
+                                        ax='time', cache=filter_cache, skip_wgt=skip_wgt, verbose=verbose, **filter_kwargs)
+                else:
+                    if 'data' not in filter_kwargs:
+                        self.pre_filter_resid[k] = self.data[k]
+                    else:
+                        self.pre_filter_resid[k] = filter_kwargs['data'][k]
+                    if 'flags' not in filter_kwargs:
+                        self.pre_filter_resid_flags[k] = self.flags[k]
+                    else:
+                        self.pre_filter_flags[k] = filter_kwargs['flags'][k]
+                if center_before_filtering:
```

Comment:
why center before filtering? shouldn't the FRP be such that it is symmetric about the main lobe anyways?

Reply:
If we want to use cleaning or DPSS fitting (to deal with missing data) which is, in some cases, only currently supports filter windows centered at zero, then we will want to center the filtering region at zero fringe-rate. In the case of DPSS fitting, non-zero window centers are supported but filters centered at zero tend to be better behaved.

Reply:
I see, so if certain modes require this than shouldn't this be hard-coded? perhaps we can for now just assume that CLEAN and DPSS only act as high-pass filters, such that this phasing should always be applied in these cases?

Reply:
I'm hardcoding this.